### PR TITLE
add distributed parallel training

### PR DESCRIPTION
**Pull request by McHaillet**
Add distributed parallel training and some other improvements to have more control over how trainers, reconstruction workers and dataloaders are divided. 

I switched to groupnorm instead of batchnorm to circumvent the normalization issues that we otherwise expected.

Time per training iteration now went from 2 hours to half an hour on the 4 gpu workstation here. With early stopping, its about 25 minutes for training per macro-iteration. I will paste my run command below to see for you if the CLI makes sense.

Also running on SHREC to see if the results remain consistent.

### CODE REVIEW

**Comment by McHaillet** [comment]
This is the run command:
```
CUDA_VISIBLE_DEVICES=0,1,2,3 OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 miss-alignment --config-file config.yaml --training-devices 0,1 --reconstruction-devices 2,2,2,3,3,3 --dataloaders-per-trainer 5
```

Training devices are given as a list, and reconstruction devices are also given as a list. 2,2,2,3,3,3 will spawn three reconstruction workers on cuda:2 and three on cuda:3 . The number of dataloaders are also specified per device that trains the model. This is because I noticed the production rate was higher than the consumption.

**Comment by dtegunov** [comment]
Ohh, the scaling factor sounds great! I'm surprised the reconstruction workers aren't keeping the GPUs busy.

Dropping BatchNorm isn't strictly necessary for distributed training, there is a wrapper in Torch that makes it work in distributed settings. Not saying we shouldn't drop it, just that it's possible.

**Comment by McHaillet** [comment]
I was a bit overoptimistic. pytorch-lightning was doing some internal dataset distribution that halved the number of steps. So I had to fix this. We are running at the correct number of steps per macro-iteration again, but that does double the training time. 

The groupnorm was also giving issues, so I had to switch it back to batchnorm but using SyncBatchNorm for multi gpu training (handled by [torch.nn.SyncBatchNorm.convert_sync_batchnorm()](https://docs.pytorch.org/docs/stable/generated/torch.nn.SyncBatchNorm.html#torch.nn.SyncBatchNorm.convert_sync_batchnorm)). I think you meant this as well. However, syncbatchnorm does reduce the speed a bit.

This is the performance difference after makes stuff work:
* before: I was running 3.70 steps/s
* after: running at 5.31 steps/s

Not as exciting as before, but its still a nice improvement.

**Comment by McHaillet** [comment]
Hmm, this turns out to not be fully solved yet. For some reason I am getting zeros in the loss after the final learning rate drop, which causes the model to immediately stop. 

I thought this was an effect of the groupnorm at first, but it also happens with synced batch norm.

**Comment by McHaillet** [comment]
Managed to get it working now finally. Apparently there was a problem in the custom MAEarlyStopping callback that didn't work with DPP. Now, it should be fine.

With SyncBatchNorm I get similar results. I am now testing with GroupNorm again because it made quite some speed difference. I'll let you know if the results remain consistent.

**Comment by McHaillet** [comment]
With syncbatchnorm after 6 macro-iterations: 
```
======================================================================
Overall Statistics
======================================================================
Average mean error across all tilt series: 1.315 Å
Std of mean errors: 0.353 Å
```

**Comment by McHaillet** [comment]
The results on SHREC with GroupNorm are also good. In the same ballpark as with SyncBatchNorm:

```
======================================================================
Overall Statistics
======================================================================
Average mean error across all tilt series: 1.135 Å
Std of mean errors: 0.272 Å
```

Full processing time for running 6 iterations (2 at 20A, 4 at 10A) : 5 hours.
On the same system with branch `main` was around 8.5 hours.
